### PR TITLE
add new mitxonline email optout table to intermediate

### PR DESCRIPTION
**Pull request by KatelynGit**
# What are the relevant tickets?
https://github.com/mitodl/hq/issues/1545

# Description (What does it do?)
adds table int__mitxonline__openedx__mysql__bulk_email_optout to intermediate

# How can this be tested?
Run the following commands against your schema or QA
Run dbt build --select intermediate.mitxonline if you have all models, otherwise add + in front of intermediate.mitxonline to run upstream models

### CODE REVIEW

**Comment by KatelynGit** [comment]
I will fix the changes unrelated to this ticket before the merge

File: src/ol_dbt/models/intermediate/mitxonline/int__mitxonline__openedx__mysql__bulk_email_optout.sql [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    from {{ ref('stg__mitxonline__app__postgres__courses_courserun') }}
+)
+
+, users as (
+    select *
+    from {{ ref('int__mitxonline__users') }}
+)
+
+select
+    email_optout.email_optout_id
```
Comment (blarghmatey) [inline src/ol_dbt/models/intermediate/mitxonline/int__mitxonline__openedx__mysql__bulk_email_optout.sql:19]: I'm curious what the utility of this field is?

**Review by blarghmatey — COMMENTED** [review]
The semantics of the email opt out table are the opposite of the opt in report that edx.org produces. In that report there is a boolean that indicates that a user has "opted in" to emails. The table that we are using is for when users "opt out" of emails. I think that for this table we may want to pull the full list of users and attach a boolean field that indicates whether the user has (implicitly) agreed to receive emails.

In other words, something like (pseudo-SQL following):
```
SELECT
 users.user_username
 , users.user_email
 , CASE WHEN email_optout.email_optout_id IS NULL THEN 1 ELSE 0 AS email_opted_in
FROM users LEFT JOIN email_optout ON email_optout.user_id = users.user_id
```

**Comment by blarghmatey** [comment]
The query used by edX to generate the opt in table is here: https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/user_api/management/commands/email_opt_in_list.py#L202-L230

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 2]
```diff
@@ -1030,6 +1052,7 @@ models:
     description: str, the full URL to the certificate on MITx Online
     tests:
     - unique
+    - not_null
```
Comment (rachellougee) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1055]: @KatelynGit Is this added by accident or any reason to change it as part of this PR? This test will fail as for revoked certificates this will be null. I removed the not_null because of that.

Comment (KatelynGit) [reply src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1055]: I can remove that. Not sure how it got added

File: src/ol_dbt/models/intermediate/mitxonline/int__mitxonline__openedx__mysql__bulk_email_optout.sql (hunk 1 above)
Comment (KatelynGit) [reply src/ol_dbt/models/intermediate/mitxonline/int__mitxonline__openedx__mysql__bulk_email_optout.sql:19]: I think the only utility of that column is basically it's use as a primary key of the table. So if we wanted to reference this table in another table it'd be easy and we'd only have to add one column. Or maybe potential use for things like indexing or one column to determine uniqueness if needed instead of multiple which might be simpler for tests.

**Comment by KatelynGit** [comment]
> The semantics of the email opt out table are the opposite of the opt in report that edx.org produces. In that report there is a boolean that indicates that a user has "opted in" to emails. The table that we are using is for when users "opt out" of emails. I think that for this table we may want to pull the full list of users and attach a boolean field that indicates whether the user has (implicitly) agreed to receive emails.
> 
> In other words, something like (pseudo-SQL following):
> 
> ```
> SELECT
>  users.user_username
>  , users.user_email
>  , CASE WHEN email_optout.email_optout_id IS NULL THEN 1 ELSE 0 AS email_opted_in
> FROM users LEFT JOIN email_optout ON email_optout.user_id = users.user_id
> ```

I've made the changes you described above. Hopefully it all looks good.

**Review by blarghmatey — APPROVED** [review]

**Comment by rachellougee** [comment]
Sorry I have to comment on this since I got notification. The table name is `int__mitxonline__openedx__mysql__bulk_email_optout` which doesn't seem to in line with our other intermediate tables. Also it's opposite to the column in this table. Am I the only one think this way?

**Comment by blarghmatey** [comment]
Thanks for calling that out Rachel. I agree that it's worth confirming to the model naming that we use for intermediate.

**Comment by KatelynGit** [comment]
> Sorry I have to comment on this since I got notification. The table name is `int__mitxonline__openedx__mysql__bulk_email_optout` which doesn't seem to in line with our other intermediate tables. Also it's opposite to the column in this table. It also doesn't have any test. Am I the only one think this way?
> 
> You can ignore this if you disagree

I've modified the table name to better align with current int naming
